RevDiff custom difftool

gerhardol opened the pull request:
Fixes #7716

Based on #8192 

## Proposed changes

Add the diff tools known to Git in a submenu in RevDiff
The list is collected asynch when the form is first created

## Screenshots 

### Before

### After

![image](https://user-images.githubusercontent.com/6248932/83955539-2afa7e00-a854-11ea-868c-0c9090612768.png)

## Test methodology 

Added a test

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
Added a fish-need-bike test, similar to most existing tests
A slightly more comprehensive test involving Git could be implemented, but I fail now (and do not want to spend more time on this).

gerhardol commented:
> A comment or `enum` regarding `mode` in `ParseCustomDiffMergeTool` would be helpful.

Added a few comments.
A regexp approach could be used when parsing too (see the issue for an example).
 
> I'd prefer a central setting, e.g. in the toolbar and/or the context menus, in favor of having a single (or few) place(s) with a submenu.
> Every place to open a difftool is of interest for me. (E.g. the history form leaves much room for improvements regarding external diffs.)

This is enough for me (a splitbutton for merging would be good too).
If I want it in some other situation, I can go and change the tool.
To change the tool would be an extra step that I would like to avoid.

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs:
```diff
@@ -171,6 +173,13 @@ private void InitializeComponent()
             // 
+            // openWithCustomDifftoolToolStripMenuItem
+            // 
+            this.openWithCustomDifftoolToolStripMenuItem.Image = global::GitUI.Properties.Images.Diff;
```
The intention here was that the submenu implicated "something more", but I will remove it

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs:
```diff
@@ -171,6 +173,13 @@ private void InitializeComponent()
+            this.openWithCustomDifftoolToolStripMenuItem.Image = global::GitUI.Properties.Images.Diff;
+            this.openWithCustomDifftoolToolStripMenuItem.Name = "openWithCustomDifftoolToolStripMenuItem";
+            this.openWithCustomDifftoolToolStripMenuItem.Size = new System.Drawing.Size(296, 22);
+            this.openWithCustomDifftoolToolStripMenuItem.Text = "Open with custom difftool";
```
OK

gerhardol replied:
First -> Second...

gerhardol commented:
> There's no need to hide easter eggs for me, e.g. in the nevertheless very helpful comment. 😉

@mstv do you always want a mention for issues you are already reviewing?

Yes settings is a little slow and custom tools are not visible (I cannot use _tortoisemerge_ but have to use a custom _tortoisemerge2_  too - should investigate). 
But it is still more often I select the commits I want to compare with a specific tool then I need to switch the default difftool in the other than "F3" diffs.

A fast switcher could be implemented in addition of course. 
There is plenty of time to change before 4.0 will ship anyway.

mergetool is only invoked from one form, so a splitbutton there is natural, no need for a fast switcher. (not in this PR anyway)

RussKie commented:
I'm itching to look at why the settings dialog is so slow, it annoys the
hell out of me

On Thu, Jul 16, 2020, 7:21 AM Gerhard Olsson <notifications@github.com>
wrote:

> There's no need to hide easter eggs for me, e.g. in the nevertheless very
> helpful comment. 😉
>
> @mstv <https://github.com/mstv> do you always want a mention for issues
> you are already reviewing?
>
> Yes settings is a little slow and custom tools are not visible (I cannot
> use *tortoisemerge* but have to use a custom *tortoisemerge2* too -
> should investigate).
> But it is still more often I select the commits I want to compare with a
> specific tool then I need to switch the default difftool in the other than
> "F3" diffs.
>
> A fast switcher could be implemented in addition of course.
> There is plenty of time to change before 4.0 will ship anyway.
>
> mergetool is only invoked from one form, so a splitbutton there is
> natural, no need for a fast switcher. (not in this PR anyway)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8194#issuecomment-659017494>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXTGNQJPUF6X6OIUVLLR3YMVDANCNFSM4NWOKWKQ>
> .
>

mstv commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -1018,6 +1018,66 @@ public static IReadOnlyList<GitItemStatus> GetSkipWorktreeFilesFromString(string
+        {
+            var tools = new List<string>();
+
+            // Simple parsing of the textual output opposite to porcelain format
+            // Explanation of the modes:
+            //  0: Initial mode (to detect first line, direct transit to next mode)
+            //  1: Built in tools, in the path (first is header, ended by a newline)
+            //  2: Newlines separating tools
+            //  3: Built in tools, in the path (first is header, ended by a newline)
+            //  4: Newlines separating tools
+            //  5: Unavailable tools, skipped
+            //  6: End note, skipped
+            //  5: Newlines
```
> > There's no need to hide easter eggs for me, e.g. in the nevertheless very helpful comment. 😉

> Do you always want a mention for issues you are already reviewing?

No, I meant just the minor inconsisties in the comment above. I didn't want to request a change.

Afterwards, I had a look at the screenshots again.

gerhardol commented:
@mstv 
I compared to the Git code for generating the list some more and adapted the parsing. It is simpler too (only one mode). Will not work if this command is translated.
There can be other approaches to the parsing, like excluding the group starting with "The following tools are valid, but not currently available:", to limit damage from output changes.
But there should be a parsable output.
If I get response from Git maintainers, I may submit a patch:
https://github.com/git/git/pull/471#issuecomment-660205096

mstv commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -1018,6 +1018,54 @@ public static IReadOnlyList<GitItemStatus> GetSkipWorktreeFilesFromString(string
+                }
+            }
+
+            return tools.OrderBy(i => i[0]).ToList();
```
Why sort only by the first character? `i` is a `string` IIUIC. So `.Sort()` should suffice.

gerhardol replied:
.Sort() on a list requires comparator, but OrderBy(i => i) works

gerhardol commented:
> The list seems to be not updated after closing FormSettings. I have waited more than the ~10 seconds it takes to fill the list after startup.

That was not implemented, not considered needed before. But a few lines of code, so I updated

gerhardol commented:
Squashed the approved commits, rebased

I tried to motivate a patch to Git to give a safer and simpler output but do not believe I can motivate such a change in Git.
The response would likely be to implement the parsing with the raw Git commands instead (but you need to access internal Git structures to get tools that are in the Git path, so those will be excluded.).
Added a few comments about that.

I also slightly changed the parsing to risk show invalid tools rather than excluding valid tools if Git output format changes. (This has not changed since the --tool-help option was added.)
Slightly simpler parsing, so I hope @mstv will approve

If approved by mstv, I would like to merge this in a couple of days
